hal/armv8m/cpu.c

```diff
@@ -0,0 +1,237 @@
+
+
+volatile cpu_context_t *_cpu_nctx;
+
+
+/* performance */
+
+
+/* TODO: add implementation */
+void hal_cpuLowPower(time_t us)
+{
+}
+
+
+void hal_cpuGetCycles(cycles_t *cb)
+{
+	/* Cycle counter is not available on armv8m
+	assumption that 1 cycle is 1us, so we use hal_timerGetUs() with 1ms resolution
+	both cycles_t and time_t have the same size on armv8m */
+	*cb = hal_timerGetUs();
+}
+
+
+void hal_cpuSetDevBusy(int s)
+{
+	spinlock_ctx_t scp;
+
+	hal_spinlockSet(&cpu_common.busySp, &scp);
+	if (s == 1) {
+		++cpu_common.busy;
+	}
+	else {
+		--cpu_common.busy;
+	}
+
+	if (cpu_common.busy < 0) {
+		cpu_common.busy = 0;
+	}
+	hal_spinlockClear(&cpu_common.busySp, &scp);
+}
+
+
+int hal_cpuCreateContext(cpu_context_t **nctx, void *start, void *kstack, size_t kstacksz, void *ustack, void *arg)
+{
+	cpu_context_t *ctx;
+
+	*nctx = 0;
+	if (kstack == NULL) {
+		return -1;
+	}
+
+	if (kstacksz < sizeof(cpu_context_t)) {
+		return -1;
+	}
+
+	/* Align user stack to 8 bytes */
+	ustack = (void *)((ptr_t)ustack & ~0x7u);
+
+	/* Prepare initial kernel stack */
+	ctx = (cpu_context_t *)(kstack + kstacksz - sizeof(cpu_context_t));
+
+	hal_memset(ctx, 0, sizeof(*ctx));
+
+	ctx->savesp_s = (u32)ctx;
+	ctx->psp_s = (ustack != NULL) ? (u32)ustack - (HWCTXSIZE * sizeof(int)) : NULL;
+	ctx->r4 = 0x44444444u;
+	ctx->r5 = 0x55555555u;
+	ctx->r6 = 0x66666666u;
+	ctx->r7 = 0x77777777u;
+	ctx->r8 = 0x88888888u;
+	ctx->r9 = 0x99999999u;
+	ctx->r10 = 0xaaaaaaaau;
+	ctx->r11 = 0xbbbbbbbbu;
+
+	if (ustack != NULL) {
+		((u32 *)ctx->psp_s)[0] = (u32)arg;    /* r0 */
+		((u32 *)ctx->psp_s)[1] = 0x11111111u; /* r1 */
+		((u32 *)ctx->psp_s)[2] = 0x22222222u; /* r2 */
+		((u32 *)ctx->psp_s)[3] = 0x33333333u; /* r3 */
+		((u32 *)ctx->psp_s)[4] = 0xccccccccu; /* r12 */
+		((u32 *)ctx->psp_s)[5] = 0xeeeeeeeeu; /* lr */
+		((u32 *)ctx->psp_s)[6] = (u32)start;  /* pc */
+		((u32 *)ctx->psp_s)[7] = 0x01000000u; /* psr */
+		ctx->irq_ret = RET_THREAD_PSP;
+	}
+	else {
+		ctx->r0 = (u32)arg;
+		ctx->r1 = 0x11111111u;
+		ctx->r2 = 0x22222222u;
+		ctx->r3 = 0x33333333u;
+		ctx->r12 = 0xccccccccu;
+		ctx->lr = 0xeeeeeeeeu;
+		ctx->pc = (u32)start;
+		ctx->psr = 0x01000000u;
+		ctx->irq_ret = RET_THREAD_MSP;
+	}
+
+	*nctx = ctx;
+	return 0;
+}
+
+
+void hal_longjmp(cpu_context_t *ctx)
+{
+	/* clang-format off */
+	__asm__ volatile(" \
+		cpsid if; \
+		str %1, [%0]; \
+		bl _hal_invokePendSV; \
+		cpsie if; \
+	1:	b 1b"
+	:
+	: "r" (&_cpu_nctx), "r" (ctx)
+	: "memory");
+	/* clang-format on */
+}
+
+
+/* core management */
+
+
+char *hal_cpuInfo(char *info)
+{
+	int i;
+	unsigned int cpuinfo;
+
+#ifdef CPU_NRF91
+	cpuinfo = _nrf91_cpuid();
+#else
+	hal_strcpy(info, "unknown");
+	return info;
+#endif
+
+	hal_strcpy(info, HAL_NAME_PLATFORM);
+	i = sizeof(HAL_NAME_PLATFORM) - 1;
+
+	if (((cpuinfo >> 24) & 0xffu) == 0x41u) {
+		hal_strcpy(info + i, "ARMv8 ");
+		i += 6;
+	}
+
+	if (((cpuinfo >> 4) & 0xfffu) == 0xc23u) {
+		hal_strcpy(info + i, "Cortex-M33 ");
+		i += 10;
+	}
```

Comment:
```suggestion
	if (((cpuinfo >> 4) & 0xfffu) == 0xD21u) {
		hal_strcpy(info + i, "Cortex-M33 ");
		i += 11;
	}
```
Update at least this.

Reply:
Oh, now I see that remote branch hasn't been updated with changes from your review @kemonats. Now they are present, thanks!